src/v/utils/file_io.cc

```diff
@@ -57,17 +58,11 @@ ss::future<> write_fully(const std::filesystem::path& p, iobuf buf) {
                  | ss::open_flags::truncate;
     /// Closes file on failure, otherwise file is expected to be closed in the
     /// success case where the ss::output_stream calls close()
-    return ss::with_file_close_on_failure(
-      ss::open_file_dma(p.string(), flags),
-      [buf = std::move(buf)](ss::file f) mutable {
-          return ss::make_file_output_stream(std::move(f), buf_size)
-            .then([buf = std::move(buf)](ss::output_stream<char> out) mutable {
-                return ss::do_with(
-                  std::move(out),
-                  [buf = std::move(buf)](ss::output_stream<char>& out) mutable {
-                      return write_iobuf_to_output_stream(std::move(buf), out)
-                        .then([&out]() mutable { return out.close(); });
-                  });
-            });
+    auto out = co_await ss::with_file_close_on_failure(
+      ss::open_file_dma(p.string(), flags), [](ss::file f) {
+          return ss::make_file_output_stream(std::move(f), buf_size);
       });
+    co_await write_iobuf_to_output_stream(std::move(buf), out).finally([&out] {
```

Comment:
question: what happens if `write_iobuf_to_output_stream` throws an exception? I'm wondering if `with_file_close_on_failure` will close the file only on failures inside the lambda passed to it or if it would close the file even on failures happening on the files later like on line 65.

question: is there any reason to use `.finally(...)` here instead of having `co_await out.close()` on a new line? `out` is captured into a lambda here, so I am wondering if `out` could get destructed before the last lambda calls `out.close()` on it.

Reply:
Q2: i want to ensure that our (and its file) is closed if 
write_iobuf_to_output_stream failed. Previously the wrapping with_file_close_on_failure would take care of that. 

Multiple ways to do it, but finally() it's one of the easiest with futures. out has the lifetime of the coroutine state, and will outlive line 67

Q1: not sure i understand the question: if write_iobuf... Throws, the finally block will run and the exception Will be propagated in the coroutine context and to the awaiter of the coroutine, like it was before. 

with_file_close_on failure is awaited immediately and has no effect after 64, if the passed future does not fail

Reply:
Thanks, that explains it.